README.md
# 1Sheeld Bootloader [![Build Status](https://travis-ci.org/Integreight/1Sheeld-Bootloader.svg?branch=master)](https://travis-ci.org/Integreight/1Sheeld-Bootloader)#

## Overview ##

1Sheeld bootloader is a small piece of firmware that come pre-installed on your 1Sheeld board and is used mainly for flashing new firmware updates to the board using the app. 

## Introduction ##

1Sheelds microcontroller's (ATmega162) flash memory is divided into two sections, the boot section and the application section. Both of them contain the bootloader and the firmware respectively. 
 
Every time the board is reset using the reset button, it will start executing the bootloader from the boot section and wait for 200ms for a request from the app to update the firmware. If the request is received, the bootloader will communicate with the app, receive the new firmware and replace the old one, it will then switch to the application section and run the new firmware. If no request is received the bootloader will exit and execute the firmware found in the application section. 
 
## Communication Method ##

The bootloader communicates with the app through serial protocol on ATmega162's UART1. 
 
## File Format ##

The program file is transferred in the binary format. So if the file is in Intel hex format (.hex) you have to transform it to binary (.bin) format before flashing using a tool like [Hex2Bin](http://hex2bin.sourceforge.net/). 
 
## Communication Protocol ##

The bootloader uses the standard [XModem](http://en.wikipedia.org/wiki/XMODEM) protocol with [CRC16](http://en.wikipedia.org/wiki/Cyclic_redundancy_check) for error checking.

Here is a breakdown of the firmware upgrading steps: 
 
- The bootloader sends a NAK to notifies the app that it is ready for the password. 
- In less than 200ms, the app sends the 8 characters password : ``` {0x64, 0x0E, 0x1C, 0x39, 0x14, 0x28, 0x57, 0xAA} ```
- The bootloader sends 'C' character to notifies the app that it is using the CRC16 method. 
- The app splits the binary file into 128 bytes packets and transmit them frame by frame in this format: 

``` 
- an SOH byte                                  {1 byte} 
- the packet number                            {1 byte} 
- the 1's complement of the packet number      {1 byte} 
- the packet                                   {128 bytes} 
- the high byte of the CRC value               {1 byte} 
- the low byte of the CRC value                {1 byte} 
```

- If the bootloader finds that the CRC values matches the computed ones, it flashs the packet and sends an ACK otherwise it sends a NAK. 
- The app waits for an ACK to transmit the next frame or a NAK to resends the last one until the whole binary file is sent. 
- After the whole binary file is sent, the app sends an EOT. 
- If there are no errors, the bootloader sends an ACK. 

## Building ##

The project is an generic C project, and can be built using any port of *avr-gcc* for either Microsoft Windows, Linux, or Mac OSX. Just make sure you have both *GNU Make* and *avr-gcc* tool chain installed on your platform then run ``` make ``` on the repo's root directory and both the debug and release versions will be built in a new subdirectory called *build*.

## Uploading ##

The ICSP pins are exposed with a 6-pin header on the bottom of your 1Sheeld board. You can easily connect any ATmega programmer and upload your own version of the bootloader.

If you are using any [USBasp programmer](http://www.fischl.de/usbasp/) and have *avrdude* installed, after building using ``` make ``` you can flash your board by running either ``` make flashdebug ``` or ``` make flashrelease ``` on the repo's root directory. Make sure you erase the chip first by running ``` make erase ```.
 
## Fuse Bits ##

- Low Value: 0xFD
- High Value: 0xD8
- Extended Value: 0xFB

Click [here](http://eleccelerator.com/fusecalc/fusecalc.php?chip=atmega162&LOW=FD&HIGH=D8&EXTENDED=FB&LOCKBIT=CC) for a description of the enabled fuse bits.

## Contribution ##

Contributions are welcomed, please follow this pattern:
- Fork the repo.
- Open an issue with your proposed feature or bug fix.
- Commit and push code to a new branch in your forked repo.
- Submit a pull request to our *development* branch.

Don't forget to drop us an email, post on our forum, or mention us on Twitter or Facebook about what you have built using 1Sheeld, we would love to hear about it.

## Changelog ##

To see what has changed in recent versions of 1Sheeld Bootloader, see the [Change Log](CHANGELOG.md).

## Note ##

1Sheeld Bootloader is based on [AVR Universal Bootloader(AVRUB)](http://sourceforge.net/projects/avrub) by Shao Ziyang.
 
## License and Copyright ##

```
This code is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License version 3 only, as
published by the Free Software Foundation.

This code is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
version 3 for more details (a copy is included in the LICENSE file that
accompanied this code).

Please contact Integreight, Inc. at info@integreight.com or post on our
support forums www.1sheeld.com/forum if you need additional information
or have any questions.
```

CHANGELOG.md
## v1.0.2 (April 10, 2016)
 - Made the repo a generic avr-gcc project that can be built with a make command.

## v1.0.1 (October 19, 2015)
 - Refactored the whole repo.
 - Removed unneeded code.
 
## v1.0.0 (March 11, 2014)
 - Initial version.

bootldr.c
/*

	Project:       1Sheeld Bootloader

	File:          bootldr.c

	Compiler:      avr-gcc 3.4.2

	Author:        Shao Ziyang (http://sourceforge.net/projects/avrub) with a little bit of modifications by Integreight, Inc. team.

	Date:          2014.5

*/

#include "bootldr.h"

int main(void)
{
	//Define some variables
	unsigned char cnt;
	unsigned char packNO;
	unsigned char crch, crcl;
	unsigned char li;
	
	//As a precaution, make SPI pins as output
	makeSPIPinsOutput();
	
	//Disable global interrupts
	cli();
	
	//Disable the watchdog timer
	disableWatchdogTimer();
	
	//Initialize Timer1 in CTC mode
	initTimer1();
	
	//Initialize UART1 with baudrate 115200
	initUART1();

	//Initialize the cnt with the password timeout count
	cnt = TIMEOUTCOUNT;
	cl = 0;
	
	//Send a starting NAK byte to notify host that we entered the bootloader section
	sendByte(XMODEM_NAK);
	
	//The 8 byte password receiving loop
	while(1)
	{
		//If Timer1 overflow, then 200ms have passed
		if(TIFR & (1<<OCF1A))   
		{
			TIFR |= (1 << OCF1A);
			
			//If we have already read 8 bytes
			if(cl == CONNECTCNT)      
			break;
			
			//Decrement the number of retries
			cnt--;
			
			//If we finished the number of retries, quit to the firmware
			if(cnt == 0)
			{
				quitToFirmwareApplication();
			}
		}

		if(isUartDataAvailable())
		{
			//Compare the key byte with its corresponding received byte
			if(readByte() == KEY[cl])
				cl++;
			else
				cl = 0;
		}
	}
	
	//Initialize the cnt with the password timeout count
	cnt = TIMEOUTCOUNTC;
	
	//Every 200ms send a "C" and wait for the first byte
	while(1)
	{
		//If Timer1 overflow, then 200ms have passed
		if(TIFR & (1 << OCF1A))  //T1 overflow
		{
			TIFR |= (1 << OCF1A);
			
			//send "C"
			sendByte(XMODEM_RWC);
			
			//Decrement the number of retries
			cnt--;
			
			//If we finished the number of retries, quit to the firmware
			if(cnt == 0)
			{
				quitToFirmwareApplication();
			}
		}
		
		if(isUartDataAvailable())
		{
			break;
		}
	}

	//Reset Timer1
	TCCR1B = 0;

	//Initialize some needed variables
	packNO    = 0;
	bufptr    = 0;
	cnt       = 0;
	flashAddress = 0;
	
	//Begin to receive data
	
	//Check if we received the end of transmission (The first time this character will be SOH)
	while(waitForTheNextByteAndReadIt() != XMODEM_EOT)
	{
		//The next expected packet number
		packNO++;
		
		//Read the packet number
		receivedPacketNumber    =  waitForTheNextByteAndReadIt();
		
		//Read the packet number complement
		packetNumberComplement	= ~waitForTheNextByteAndReadIt();
		
		//Receive a full data frame
		for(li = 0; li < BUFFERSIZE; li++)
		{
			buf[bufptr] = waitForTheNextByteAndReadIt();
			bufptr++;
		}
		
		//Read the CRC bytes
		crch = waitForTheNextByteAndReadIt();
		crcl = waitForTheNextByteAndReadIt();
		
		//Check if the received packet number and its complement matches the expected values
		if ((packNO == receivedPacketNumber) && (packNO == packetNumberComplement))
		{
			//Calculate the CRC checksum
			calculateCRCChecksum(&buf[bufptr - BUFFERSIZE]);
			
			//Check if it is the same as the received ones
			if((crch == ch) && (crcl == cl))
			{
				//Avoid writing to the boot section
				if(flashAddress < BOOTSTARTADDRESS)
				{
					//If flash page is full
					if(bufptr >= SPM_PAGESIZE)
					{
						//Write the flash page
						writeOneFlashPage(buf);
						
						//Modify the flash page address
						flashAddress += SPM_PAGESIZE;
						
						//Reset the received data buffer pointer
						bufptr = 0;
					}
				}
				
				//Ignore flash write when flash address exceeds BootStart
				else                                  
				{
					//Reset the received data buffer pointer
					bufptr = 0;
				}
				
				//Read flash and compare it with the buffer's content
				if(flashAddress < BOOTSTARTADDRESS)
				{
					//Enable the application section
					boot_rww_enable();
					
					//Clear the error flag
					cl = 1;
					
					for(pagptr = 0; pagptr < BUFFERSIZE; pagptr++)
					{
						if(pgm_read_byte(flashAddress - BUFFERSIZE + pagptr) != buf[pagptr])
						{
							//Set the error flag and break from the loop
							cl = 0;                         
							break;
						}
					}
					
					//If CRC bytes are equal, send ACK
					if(cl)
					{
						sendByte(XMODEM_ACK);
						cnt = 0;
					}
					
					//CRC checksum error, tell the host to resend
					else
					{
						//Return the packet number to the expected value
						packNO--;
						
						//Reset the received data buffer pointer
						bufptr = 0;
						
						//Send NAK byte
						sendByte(XMODEM_NAK);
						
						//Increase the error count
						cnt++;
						
						//Modify the flash page address
						flashAddress -= BUFFERSIZE;
					}
				}
				else                                  //don't need verify, send ACK directly
				{
					sendByte(XMODEM_ACK);
					cnt = 0;
				}
			}
			
			//The CRC values are not the expected ones, tell the host to resend
			else
			{
				//Return the packet number to the expected value
				packNO--;
				
				//Reset the received data buffer pointer
				bufptr = 0;
				
				//Send NAK byte
				sendByte(XMODEM_NAK);
				
				//Increase the error count
				cnt++;
			}
		}
		
		//The received packet number and its complement don't match the expected values
		else
		{
			//Return the packet number to the expected value
			packNO--;
			
			//Reset the received data buffer pointer
			bufptr = 0;
			
			//Send NAK byte
			sendByte(XMODEM_NAK);
			
			//Increase the error count
			cnt++;
		}
		
		//If there is too many errors, abort
		if(cnt > 3)
		{
			break;
		}
		
	}
	
	//If there is no errors, send ACK
	if(cnt == 0)
	{
		sendByte(XMODEM_ACK);
	}
	else
	{
		sendByte(XMODEM_CAN);
	}
	
	//Go to the firmware all times
	quitToFirmwareApplication();
	return 0;
}

//End of file: bootldr.c


bootldr.h
/*

	Project:       1Sheeld Bootloader

	File:          bootldr.h

	Compiler:      avr-gcc 3.4.2

	Author:        Shao Ziyang (http://sourceforge.net/projects/avrub) with a little bit of modifications by Integreight, Inc. team.

	Date:          2014.5

*/

#ifndef _BOOTLDR_H_
#define _BOOTLDR_H_

#include <avr/io.h>
#include <avr/wdt.h>
#include <avr/boot.h>
#include <avr/pgmspace.h>
#include <avr/interrupt.h>

//Define UART buffer's length
#define BUFFERSIZE			128

//Define system clock
#define F_CPU				7372800UL

//Define communication's baudrate
#define BAUDRATE			115200

//Boot section start address(byte)
#define BOOTSTARTADDRESS	0x1C00 * 2

//Basic timer interval in ms
#define TIMERINTERVAL		200

//Define the password waiting timout
//(Maximum password waiting time = TIMEOUTCOUNT * TIMERINTERVAL)
#define TIMEOUTCOUNT		1

//Define the data waiting timout
//(Maximum data waiting time = TIMEOUTCOUNTC * TIMERINTERVAL)
#define TIMEOUTCOUNTC		250

//Password character length
#define CONNECTCNT			8

//The password
unsigned char KEY[] = {0x64, 0x0E, 0x1C, 0x39, 0x14, 0x28, 0x57, 0xAA};

//The firmware application start address
#define PROG_START         0x0000

//Used X-modem control commands
#define XMODEM_EOT         0x04
#define XMODEM_ACK         0x06
#define XMODEM_NAK         0x15
#define XMODEM_CAN         0x18
#define XMODEM_RWC         'C'
#define isUartDataAvailable()        (UCSR1A & (1 << RXC1))
#define readByte()          UDR1

//Timer1 configuration: (prescaler: 1024, CTC mode: 4 and interval unit is millisecond)
#define initTimer1()															\
{																			\
	OCR1A  = (unsigned int)(TIMERINTERVAL * (F_CPU  / (1024 * 1000.0f)));	\
	TCCR1B = (1 << WGM12)|(1 << CS12)|(1 << CS10);							\
}

//Configure SPI pins as output as a precaution
#define makeSPIPinsOutput()						\
{												\
	DDRB|=(1<<DDB5)|(1<<DDB6)|(1<<DDB7);		\
}

//Disable watchdog timer
#define disableWatchdogTimer()					\
{												\
	MCUCSR = 0;									\
	wdt_disable();								\
}

//Define UART1 UBBR value
#define UBRR_BOOT_VALUE 3

//Initialize UART1
#define initUART1()												\
{                                                               \
	UBRR1L = UBRR_BOOT_VALUE;                                   \
	UCSR1C = (1<<URSEL1) | (1 << UCSZ10)|(1 << UCSZ11);			\
	UCSR1B = (1 << RXEN1) | (1 << TXEN1);						\
}

//Define a receive buffer and some variables
unsigned char buf[BUFFERSIZE];
unsigned char bufptr, pagptr;
unsigned char ch, cl, receivedPacketNumber, packetNumberComplement;
unsigned int flashAddress;

//Write one flash page
void writeOneFlashPage(unsigned char *buf)
{
	//Erase one flash page
	boot_page_erase(flashAddress);
	//Wait till the flash page is erased
	boot_spm_busy_wait();
	//Fill data to the flash buffer
	for(pagptr = 0; pagptr < SPM_PAGESIZE; pagptr += 2)
	{
		boot_page_fill(pagptr, buf[pagptr] + (buf[pagptr + 1] << 8));
	}
	//Write buffer to one flash page
	boot_page_write(flashAddress);
	//Wait till the flash page is written
	boot_spm_busy_wait();
}

//Quit the bootloader and jump to the firmware application
void quitToFirmwareApplication()
{
	//Reset Timer1 register
	TCCR1B = 0;
	//Enable the application section
	boot_rww_enable();
	//Jump to the firmware
	(*((void(*)(void))PROG_START))();
}

//Send a byte to the UART1
void sendByte(unsigned char data)
{
	UDR1 = data;
	//Wait for the sending to finish
	while(!(UCSR1A & (1<<TXC1)));
	UCSR1A |= (1 << TXC1);
}

//Wait to receive a byte from UART1
unsigned char waitForTheNextByteAndReadIt()
{
	while(!isUartDataAvailable());
	return readByte();
}

//Calculate CRC checksum
void calculateCRCChecksum(unsigned char *buf)
{
	unsigned char j;
	unsigned char i;
	unsigned int t;
	unsigned int crc;
	crc = 0;
	for(j = BUFFERSIZE; j > 0; j--)
	{
		//CRC1021 checksum
		crc = (crc ^ (((unsigned int) *buf) << 8));
		for(i = 8; i > 0; i--)
		{
			t = crc << 1;
			if(crc & 0x8000)
			t = t ^ 0x1021;
			crc = t;
		}
		buf++;
	}
	ch = crc / 256;
	cl = crc % 256;
}

#endif

//End of file: bootldr.h


bootloader.c
/*

	Project:       1Sheeld Bootloader

	File:          bootldr.c

	Compiler:      avr-gcc 3.4.2

	Author:        Shao Ziyang (http://sourceforge.net/projects/avrub) with a little bit of modifications by Integreight, Inc. team.

	Date:          2014.5

*/

#include "bootloader.h"

int main(void)
{
	//Define some variables
	unsigned char cnt;
	unsigned char packNO;
	unsigned char crch, crcl;
	unsigned char li;
	
	//As a precaution, make SPI pins as output
	makeSPIPinsOutput();
	
	//Disable global interrupts
	cli();
	
	//Disable the watchdog timer
	disableWatchdogTimer();
	
	//Initialize Timer1 in CTC mode
	initTimer1();
	
	//Initialize UART1 with baudrate 115200
	initUART1();

	//Initialize the cnt with the password timeout count
	cnt = TIMEOUTCOUNT;
	cl = 0;
	
	//Send a starting NAK byte to notify host that we entered the bootloader section
	sendByte(XMODEM_NAK);
	
	//The 8 byte password receiving loop
	while(1)
	{
		//If Timer1 overflow, then 200ms have passed
		if(TIFR & (1<<OCF1A))   
		{
			TIFR |= (1 << OCF1A);
			
			//If we have already read 8 bytes
			if(cl == CONNECTCNT)      
			break;
			
			//Decrement the number of retries
			cnt--;
			
			//If we finished the number of retries, quit to the firmware
			if(cnt == 0)
			{
				quitToFirmwareApplication();
			}
		}

		if(isUartDataAvailable())
		{
			//Compare the key byte with its corresponding received byte
			if(readByte() == KEY[cl])
				cl++;
			else
				cl = 0;
		}
	}
	
	//Initialize the cnt with the password timeout count
	cnt = TIMEOUTCOUNTC;
	
	//Every 200ms send a "C" and wait for the first byte
	while(1)
	{
		//If Timer1 overflow, then 200ms have passed
		if(TIFR & (1 << OCF1A))  //T1 overflow
		{
			TIFR |= (1 << OCF1A);
			
			//send "C"
			sendByte(XMODEM_RWC);
			
			//Decrement the number of retries
			cnt--;
			
			//If we finished the number of retries, quit to the firmware
			if(cnt == 0)
			{
				quitToFirmwareApplication();
			}
		}
		
		if(isUartDataAvailable())
		{
			break;
		}
	}

	//Reset Timer1
	TCCR1B = 0;

	//Initialize some needed variables
	packNO    = 0;
	bufptr    = 0;
	cnt       = 0;
	flashAddress = 0;
	
	//Begin to receive data
	
	//Check if we received the end of transmission (The first time this character will be SOH)
	while(waitForTheNextByteAndReadIt() != XMODEM_EOT)
	{
		//The next expected packet number
		packNO++;
		
		//Read the packet number
		receivedPacketNumber    =  waitForTheNextByteAndReadIt();
		
		//Read the packet number complement
		packetNumberComplement	= ~waitForTheNextByteAndReadIt();
		
		//Receive a full data frame
		for(li = 0; li < BUFFERSIZE; li++)
		{
			buf[bufptr] = waitForTheNextByteAndReadIt();
			bufptr++;
		}
		
		//Read the CRC bytes
		crch = waitForTheNextByteAndReadIt();
		crcl = waitForTheNextByteAndReadIt();
		
		//Check if the received packet number and its complement matches the expected values
		if ((packNO == receivedPacketNumber) && (packNO == packetNumberComplement))
		{
			//Calculate the CRC checksum
			calculateCRCChecksum(&buf[bufptr - BUFFERSIZE]);
			
			//Check if it is the same as the received ones
			if((crch == ch) && (crcl == cl))
			{
				//Avoid writing to the boot section
				if(flashAddress < BOOTSTARTADDRESS)
				{
					//If flash page is full
					if(bufptr >= SPM_PAGESIZE)
					{
						//Write the flash page
						writeOneFlashPage(buf);
						
						//Modify the flash page address
						flashAddress += SPM_PAGESIZE;
						
						//Reset the received data buffer pointer
						bufptr = 0;
					}
				}
				
				//Ignore flash write when flash address exceeds BootStart
				else                                  
				{
					//Reset the received data buffer pointer
					bufptr = 0;
				}
				
				//Read flash and compare it with the buffer's content
				if(flashAddress < BOOTSTARTADDRESS)
				{
					//Enable the application section
					boot_rww_enable();
					
					//Clear the error flag
					cl = 1;
					
					for(pagptr = 0; pagptr < BUFFERSIZE; pagptr++)
					{
						if(pgm_read_byte(flashAddress - BUFFERSIZE + pagptr) != buf[pagptr])
						{
							//Set the error flag and break from the loop
							cl = 0;                         
							break;
						}
					}
					
					//If CRC bytes are equal, send ACK
					if(cl)
					{
						sendByte(XMODEM_ACK);
						cnt = 0;
					}
					
					//CRC checksum error, tell the host to resend
					else
					{
						//Return the packet number to the expected value
						packNO--;
						
						//Reset the received data buffer pointer
						bufptr = 0;
						
						//Send NAK byte
						sendByte(XMODEM_NAK);
						
						//Increase the error count
						cnt++;
						
						//Modify the flash page address
						flashAddress -= BUFFERSIZE;
					}
				}
				else                                  //don't need verify, send ACK directly
				{
					sendByte(XMODEM_ACK);
					cnt = 0;
				}
			}
			
			//The CRC values are not the expected ones, tell the host to resend
			else
			{
				//Return the packet number to the expected value
				packNO--;
				
				//Reset the received data buffer pointer
				bufptr = 0;
				
				//Send NAK byte
				sendByte(XMODEM_NAK);
				
				//Increase the error count
				cnt++;
			}
		}
		
		//The received packet number and its complement don't match the expected values
		else
		{
			//Return the packet number to the expected value
			packNO--;
			
			//Reset the received data buffer pointer
			bufptr = 0;
			
			//Send NAK byte
			sendByte(XMODEM_NAK);
			
			//Increase the error count
			cnt++;
		}
		
		//If there is too many errors, abort
		if(cnt > 3)
		{
			break;
		}
		
	}
	
	//If there is no errors, send ACK
	if(cnt == 0)
	{
		sendByte(XMODEM_ACK);
	}
	else
	{
		sendByte(XMODEM_CAN);
	}
	
	//Go to the firmware all times
	quitToFirmwareApplication();
	return 0;
}

//End of file: boorloader.c


1Sheeld Bootloader/bootcfg.h
/*

  Project:       1Sheeld Bootloader
  
  File:          bootcfg.h
                
  Compiler:      avr-gcc 3.4.2

  Author:        Integreight based on AVR Universal Bootloader by Shao ziyang: http://sourceforge.net/projects/avrub
                 
  Date:          2014.5

*/

#ifndef _BOOTCFG_H_
#define _BOOTCFG_H_        1

////define uart buffer's length
//#define BUFFERSIZE         128
//
////system clock(Hz)
//#ifndef F_CPU
//#define F_CPU              7372800UL
//#endif
//
////baudrate
//#define BAUDRATE           115200
//
////Boot section start address(byte)
////define BootStart to 0 will disable this function
//#define BOOTSTARTADDRESS          0x1C00 * 2

//verify flash's data while write
//ChipCheck will only take effect while BootStart enable also
//#define VERIFYFLASH          1

//Bootloader launch  0:comport password  1:port level
//#define LEVELMODE          0

//#define LEVELPORT          D
//#define LEVELPIN           PD7
//port level  1:High  0:Low
//#define PINLEVEL           0

////max wait password time = TimeOutCnt * timeclk
////timeout count
//#define TIMEOUTCOUNT         1
//
////basic timer interval(ms)
//#define TIMERINTERVAL        200
//
////max wait data time = TimeOutCntC * timeclk
////send 'C' command count
//#define TIMEOUTCOUNTC        250
//
////password length
//#define CONNECTCNT         8
//
////password
////#if LEVELMODE == 0
//unsigned char KEY[] = {0x64, 0x0E, 0x1C, 0x39, 0x14, 0x28, 0x57, 0xAA};  
////#endif
//
////comport number: 0/1/2..
//#define COMPORTNO          1

////enable watchdog
//#define WDGEN              0

////enable RS485/RS422 mode
//#define RS485              0
////RS485/RS422 send control port
//#define RS485PORT          D
//#define RS485TXEn          PD5 

////enable LED indication
//#define LEDEN              1
//LED control port
//#define LEDPORT            D
//#define LEDPORTNO          PD6

////some kind of AVR need special delay after comport initialization
//#define INITUARTDELAY      0 

////communication checksum method   0:CRC16  1:add up
//#define CRCMODE            0
//
////Verbose mode: display more prompt message
//#define VERBOSE            0

#endif

//End of file: bootcfg.h


bootloader.h
/*

	Project:       1Sheeld Bootloader

	File:          bootldr.h

	Compiler:      avr-gcc 3.4.2

	Author:        Shao Ziyang (http://sourceforge.net/projects/avrub) with a little bit of modifications by Integreight, Inc. team.

	Date:          2014.5

*/

#ifndef _BOOTLOADER_H_
#define _BOOTLOADER_H_

#include <avr/io.h>
#include <avr/wdt.h>
#include <avr/boot.h>
#include <avr/pgmspace.h>
#include <avr/interrupt.h>

//Define UART buffer's length
#define BUFFERSIZE			128

//Define system clock
#define F_CPU				7372800UL

//Define communication's baudrate
#define BAUDRATE			115200

//Boot section start address(byte)
#define BOOTSTARTADDRESS	0x1C00 * 2

//Basic timer interval in ms
#define TIMERINTERVAL		200

//Define the password waiting timout
//(Maximum password waiting time = TIMEOUTCOUNT * TIMERINTERVAL)
#define TIMEOUTCOUNT		1

//Define the data waiting timout
//(Maximum data waiting time = TIMEOUTCOUNTC * TIMERINTERVAL)
#define TIMEOUTCOUNTC		250

//Password character length
#define CONNECTCNT			8

//The password
unsigned char KEY[] = {0x64, 0x0E, 0x1C, 0x39, 0x14, 0x28, 0x57, 0xAA};

//The firmware application start address
#define PROG_START         0x0000

//Used X-modem control commands
#define XMODEM_EOT         0x04
#define XMODEM_ACK         0x06
#define XMODEM_NAK         0x15
#define XMODEM_CAN         0x18
#define XMODEM_RWC         'C'
#define isUartDataAvailable()        (UCSR1A & (1 << RXC1))
#define readByte()          UDR1

//Timer1 configuration: (prescaler: 1024, CTC mode: 4 and interval unit is millisecond)
#define initTimer1()															\
{																			\
	OCR1A  = (unsigned int)(TIMERINTERVAL * (F_CPU  / (1024 * 1000.0f)));	\
	TCCR1B = (1 << WGM12)|(1 << CS12)|(1 << CS10);							\
}

//Configure SPI pins as output as a precaution
#define makeSPIPinsOutput()						\
{												\
	DDRB|=(1<<DDB5)|(1<<DDB6)|(1<<DDB7);		\
}

//Disable watchdog timer
#define disableWatchdogTimer()					\
{												\
	MCUCSR = 0;									\
	wdt_disable();								\
}

//Define UART1 UBBR value
#define UBRR_BOOT_VALUE 3

//Initialize UART1
#define initUART1()												\
{                                                               \
	UBRR1L = UBRR_BOOT_VALUE;                                   \
	UCSR1C = (1<<URSEL1) | (1 << UCSZ10)|(1 << UCSZ11);			\
	UCSR1B = (1 << RXEN1) | (1 << TXEN1);						\
}

//Define a receive buffer and some variables
unsigned char buf[BUFFERSIZE];
unsigned char bufptr, pagptr;
unsigned char ch, cl, receivedPacketNumber, packetNumberComplement;
unsigned int flashAddress;

//Write one flash page
void writeOneFlashPage(unsigned char *buf)
{
	//Erase one flash page
	boot_page_erase(flashAddress);
	//Wait till the flash page is erased
	boot_spm_busy_wait();
	//Fill data to the flash buffer
	for(pagptr = 0; pagptr < SPM_PAGESIZE; pagptr += 2)
	{
		boot_page_fill(pagptr, buf[pagptr] + (buf[pagptr + 1] << 8));
	}
	//Write buffer to one flash page
	boot_page_write(flashAddress);
	//Wait till the flash page is written
	boot_spm_busy_wait();
}

//Quit the bootloader and jump to the firmware application
void quitToFirmwareApplication()
{
	//Reset Timer1 register
	TCCR1B = 0;
	//Enable the application section
	boot_rww_enable();
	//Jump to the firmware
	(*((void(*)(void))PROG_START))();
}

//Send a byte to the UART1
void sendByte(unsigned char data)
{
	UDR1 = data;
	//Wait for the sending to finish
	while(!(UCSR1A & (1<<TXC1)));
	UCSR1A |= (1 << TXC1);
}

//Wait to receive a byte from UART1
unsigned char waitForTheNextByteAndReadIt()
{
	while(!isUartDataAvailable());
	return readByte();
}

//Calculate CRC checksum
void calculateCRCChecksum(unsigned char *buf)
{
	unsigned char j;
	unsigned char i;
	unsigned int t;
	unsigned int crc;
	crc = 0;
	for(j = BUFFERSIZE; j > 0; j--)
	{
		//CRC1021 checksum
		crc = (crc ^ (((unsigned int) *buf) << 8));
		for(i = 8; i > 0; i--)
		{
			t = crc << 1;
			if(crc & 0x8000)
			t = t ^ 0x1021;
			crc = t;
		}
		buf++;
	}
	ch = crc / 256;
	cl = crc % 256;
}

#endif

//End of file: bootloader.h
